src/commands/deployment/correlate.ts

```diff
@@ -0,0 +1,182 @@
+import {getCISpanTags} from '../../helpers/ci'
+import {gitRepositoryURL, gitLocalCommitShas, gitCurrentBranch} from '../../helpers/git/get-git-data'
+import {Logger, LogLevel} from '../../helpers/logger'
+import {retryRequest} from '../../helpers/retry'
+import {CI_PROVIDER_NAME, CI_ENV_VARS, GIT_REPOSITORY_URL, GIT_SHA} from '../../helpers/tags'
+import {getApiHostForSite, getRequestBuilder} from '../../helpers/utils'
+
+/**
+ * This command is a wrapper around the datadog-ci tag command, allowing customers to mark CI jobs
+ * as deployments and setting specific properties, like the environment or the revision in a simple way.
+ */
+export class DeploymentCorrelateCommand extends Command {
+  public static paths = [['deployment', 'correlate']]
+
+  public static usage = Command.Usage({
+    category: 'CI Visibility',
+    description: 'Correlate gitOps CD deployments with CI pipelines.',
+    details: `
+      This command will correlate the pipeline with a gitOps CD deployment.\n
+      It does not work for every setup, see README for details.
+    `,
+    examples: [['Correlate an argoCD deployment', 'datadog-ci deployment correlate --provider argocd']],
+  })
+
+  private cdProviderParam = Option.String('--provider')
+  private cdProvider!: string
+  private supportedCDProviders: string[] = ['argocd'] // update the README if changing this
+  private configurationRepo = Option.String('--config-repo')
+  private dryRun = Option.Boolean('--dry-run', false)
+
+  private config = {
+    apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
+  }
+
+  private logger: Logger = new Logger((s: string) => this.context.stdout.write(s), LogLevel.INFO)
+
+  public async execute() {
+    if (!this.config.apiKey) {
+      this.logger.error(
+        `Neither ${chalk.red.bold('DATADOG_API_KEY')} nor ${chalk.red.bold('DD_API_KEY')} is in your environment.`
+      )
+
+      return 1
+    }
+
+    if (!this.validateCDProviderParam()) {
+      return 1
+    }
+
+    const tags = getCISpanTags() || {}
+
+    if (!this.validateTags(tags)) {
+      return 1
+    }
+
+    let envVars: Record<string, string> = {}
+    if (tags[CI_ENV_VARS]) {
+      envVars = JSON.parse(tags[CI_ENV_VARS])
+      delete tags[CI_ENV_VARS]
+    }
+    const ciEnv: Record<string, string> = {
+      ...tags,
+      ...envVars,
+    }
+
+    const git = simpleGit({
+      baseDir: process.cwd(),
+      binary: 'git',
+      maxConcurrentProcesses: 2, // max 2 git commands at the same time
+    })
+
+    const currentBranch = await gitCurrentBranch(git)
+    if (!currentBranch) {
+      this.logger.error('Could not get current branch')
+
+      return 1
+    }
+
+    let localCommitShas: string[]
+    if (this.configurationRepo) {
+      localCommitShas = await gitLocalCommitShas(git, currentBranch)
+    } else {
+      ;[this.configurationRepo, localCommitShas] = await Promise.all([
+        gitRepositoryURL(git),
+        gitLocalCommitShas(git, currentBranch),
+      ])
+    }
```

Comment:
Is this because we want to parallelize the git calls? How much do they take? If they do not take a lot, I would prefer to have this sequentially and make the logic simpler

Reply:
Yeah I would rather parallelise them. It will not make  the CLI much faster but it also does not add much complexity into the command code IMO